### ISSUE TITLE
data-loading-infrastructure

### ISSUE THREAD

**Issue by louiswang524** [louiswang524/pytorch-recsys#4]
# Task: Data Loading Infrastructure with Basic Preprocessing

## Objective
Implement a robust and efficient data loading infrastructure that can handle sequential recommendation datasets with support for implicit/explicit feedback, flexible preprocessing, and optimized batch loading for long sequences.

## Background
Sequential recommendation models require specialized data handling for temporal sequences, padding, negative sampling, and efficient batching. This infrastructure must support sequences up to 1000 items while maintaining memory efficiency and training speed.

### Core Data Components
- **Sequential Dataset**: Handle user interaction sequences with temporal ordering
- **Data Loaders**: Efficient batch loading with proper collation for variable-length sequences
- **Preprocessing Pipeline**: Data cleaning, filtering, and transformation utilities
- **Negative Sampling**: Multiple negative sampling strategies for implicit feedback

### Dataset Support
- **Format Support**: CSV, JSON, Parquet for flexibility
- **Standard Datasets**: MovieLens, Amazon, with automatic download and preprocessing
- **Custom Datasets**: Easy integration of proprietary datasets
- **Data Validation**: Schema validation and quality checks

### Sequential Dataset Implementation
```python
from typing import Dict, List, Optional, Tuple, Union, Any
import torch
from torch.utils.data import Dataset
import pandas as pd
import numpy as np
from pathlib import Path
from omegaconf import DictConfig

class SequentialDataset(Dataset):
    """Dataset for sequential recommendation models"""
    
    def __init__(
        self,
        data_path: str,
        config: DictConfig,
        split: str = 'train',
        transform: Optional[callable] = None
    ):
        super().__init__()
        self.data_path = Path(data_path)
        self.config = config
        self.split = split
        self.transform = transform
        
        # Configuration parameters
        self.max_seq_len = config.max_seq_len
        self.min_seq_len = config.get('min_seq_len', 2)
        self.negative_sampling = config.get('negative_sampling', True)
        self.num_negatives = config.get('num_negatives', 1)
        
        # Load and preprocess data
        self.sequences, self.metadata = self._load_data()
        self.item_vocab, self.user_vocab = self._build_vocabularies()
        self.sequences_encoded = self._encode_sequences()
        
        print(f"Loaded {len(self.sequences_encoded)} sequences for {split}")
    
    def __len__(self) -> int:
        return len(self.sequences_encoded)
    
    def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
        """Get sequence sample with target and negatives"""
        sequence = self.sequences_encoded[idx]
        
        # Handle sequence length
        if len(sequence) > self.max_seq_len + 1:  # +1 for target
            # Random crop for long sequences
            start_idx = np.random.randint(0, len(sequence) - self.max_seq_len)
            sequence = sequence[start_idx:start_idx + self.max_seq_len + 1]
        
        # Split sequence and target
        if len(sequence) > 1:
            seq_items = sequence[:-1]
            target_item = sequence[-1]
        else:
            seq_items = [0]  # Padding token
            target_item = sequence[0]
        
        # Pad sequence to max length
        seq_items = self._pad_sequence(seq_items)
        
        sample = {
            'sequences': torch.tensor(seq_items, dtype=torch.long),
            'targets': torch.tensor(target_item, dtype=torch.long),
            'seq_lengths': torch.tensor(len(seq_items), dtype=torch.long)
        }
        
        # Add negative samples if needed
        if self.negative_sampling and self.split == 'train':
            negatives = self._sample_negatives(target_item)
            sample['negatives'] = torch.tensor(negatives, dtype=torch.long)
        
        # Apply transforms
        if self.transform:
            sample = self.transform(sample)
        
        return sample
    
    def _load_data(self) -> Tuple[List[List[int]], Dict[str, Any]]:
        """Load raw data from file"""
        file_path = self.data_path / f"{self.split}.csv"
        
        if not file_path.exists():
            raise FileNotFoundError(f"Data file not found: {file_path}")
        
        # Load data
        df = pd.read_csv(file_path)
        
        # Validate required columns
        required_cols = ['user_id', 'item_id', 'timestamp']
        missing_cols = [col for col in required_cols if col not in df.columns]
        if missing_cols:
            raise ValueError(f"Missing required columns: {missing_cols}")
        
        # Sort by user and timestamp
        df = df.sort_values(['user_id', 'timestamp'])
        
        # Group by user to create sequences
        sequences = []
        user_stats = []
        
        for user_id, user_data in df.groupby('user_id'):
            items = user_data['item_id'].tolist()
            
            # Filter by minimum sequence length
            if len(items) >= self.min_seq_len:
                sequences.append(items)
                user_stats.append({
                    'user_id': user_id,
                    'num_interactions': len(items)
                })
        
        metadata = {
            'num_users': len(user_stats),
            'num_sequences': len(sequences),
            'avg_seq_length': np.mean([len(seq) for seq in sequences]),
            'user_stats': user_stats
        }
        
        return sequences, metadata
    
    def _build_vocabularies(self) -> Tuple[Dict[int, int], Dict[int, int]]:
        """Build item and user vocabularies"""
        # Get all unique items
        all_items = set()
        for sequence in self.sequences:
            all_items.update(sequence)
        
        # Create item vocabulary (0 reserved for padding)
        item_vocab = {item_id: idx + 1 for idx, item_id in enumerate(sorted(all_items))}
        item_vocab[0] = 0  # Padding token
        
        # User vocabulary (if needed)
        user_vocab = {}  # Can be implemented later if needed
        
        return item_vocab, user_vocab
    
    def _encode_sequences(self) -> List[List[int]]:
        """Encode sequences using vocabulary"""
        encoded = []
        for sequence in self.sequences:
            encoded_seq = [self.item_vocab[item_id] for item_id in sequence]
            encoded.append(encoded_seq)
        return encoded
    
    def _pad_sequence(self, sequence: List[int]) -> List[int]:
        """Pad sequence to max length"""
        if len(sequence) >= self.max_seq_len:
            return sequence[-self.max_seq_len:]  # Take last max_seq_len items
        else:
            # Pad with zeros at the beginning
            padding = [0] * (self.max_seq_len - len(sequence))
            return padding + sequence
    
    def _sample_negatives(self, target_item: int) -> List[int]:
        """Sample negative items for training"""
        num_items = len(self.item_vocab)
        negatives = []
        
        while len(negatives) < self.num_negatives:
            # Random sampling from all items
            neg_item = np.random.randint(1, num_items)  # Skip padding token
            if neg_item != target_item:
                negatives.append(neg_item)
        
        return negatives
    
    def get_stats(self) -> Dict[str, Any]:
        """Get dataset statistics"""
        seq_lengths = [len(seq) for seq in self.sequences]
        
        stats = {
            'num_sequences': len(self.sequences),
            'num_items': len(self.item_vocab) - 1,  # Exclude padding
            'avg_seq_length': np.mean(seq_lengths),
            'min_seq_length': np.min(seq_lengths),
            'max_seq_length': np.max(seq_lengths),
            'std_seq_length': np.std(seq_lengths),
        }
        
        return stats
```

### Data Collation Function
```python
from torch.utils.data import DataLoader
from torch.nn.utils.rnn import pad_sequence

def collate_fn(batch: List[Dict[str, torch.Tensor]]) -> Dict[str, torch.Tensor]:
    """Custom collation function for sequential data"""
    
    # Extract components
    sequences = [item['sequences'] for item in batch]
    targets = [item['targets'] for item in batch]
    seq_lengths = [item['seq_lengths'] for item in batch]
    
    # Stack sequences and targets
    sequences_tensor = torch.stack(sequences)
    targets_tensor = torch.stack(targets)
    seq_lengths_tensor = torch.stack(seq_lengths)
    
    collated = {
        'sequences': sequences_tensor,
        'targets': targets_tensor,
        'seq_lengths': seq_lengths_tensor
    }
    
    # Handle negatives if present
    if 'negatives' in batch[0]:
        negatives = [item['negatives'] for item in batch]
        collated['negatives'] = torch.stack(negatives)
    
    return collated

def create_dataloader(
    dataset: SequentialDataset,
    batch_size: int,
    shuffle: bool = True,
    num_workers: int = 4
) -> DataLoader:
    """Create optimized DataLoader for sequential datasets"""
    
    return DataLoader(
        dataset,
        batch_size=batch_size,
        shuffle=shuffle,
        num_workers=num_workers,
        collate_fn=collate_fn,
        pin_memory=torch.cuda.is_available(),
        persistent_workers=True if num_workers > 0 else False
    )
```

### Standard Dataset Implementations
```python
import urllib.request
import zipfile
from pathlib import Path

class MovieLensDataset(SequentialDataset):
    """MovieLens dataset with automatic download"""
    
    URLS = {
        'ml-100k': 'http://files.grouplens.org/datasets/movielens/ml-100k.zip',
        'ml-1m': 'http://files.grouplens.org/datasets/movielens/ml-1m.zip',
        'ml-10m': 'http://files.grouplens.org/datasets/movielens/ml-10m.zip'
    }
    
    def __init__(self, root: str, dataset: str = 'ml-1m', **kwargs):
        self.root = Path(root)
        self.dataset = dataset
        
        # Download and prepare data
        self._download_and_prepare()
        
        # Initialize parent class
        data_path = self.root / dataset / 'processed'
        super().__init__(data_path, **kwargs)
    
    def _download_and_prepare(self):
        """Download and prepare MovieLens data"""
        dataset_dir = self.root / self.dataset
        processed_dir = dataset_dir / 'processed'
        
        if processed_dir.exists():
            return  # Already processed
        
        # Download if needed
        raw_dir = dataset_dir / 'raw'
        if not raw_dir.exists():
            self._download(raw_dir)
        
        # Process data
        self._process_data(raw_dir, processed_dir)
    
    def _download(self, raw_dir: Path):
        """Download MovieLens dataset"""
        raw_dir.mkdir(parents=True, exist_ok=True)
        
        url = self.URLS[self.dataset]
        zip_path = raw_dir / f"{self.dataset}.zip"
        
        print(f"Downloading {self.dataset}...")
        urllib.request.urlretrieve(url, zip_path)
        
        # Extract
        with zipfile.ZipFile(zip_path, 'r') as zip_ref:
            zip_ref.extractall(raw_dir)
    
    def _process_data(self, raw_dir: Path, processed_dir: Path):
        """Process raw MovieLens data to sequential format"""
        processed_dir.mkdir(parents=True, exist_ok=True)
        
        # Load ratings data
        if self.dataset == 'ml-100k':
            ratings_file = raw_dir / 'ml-100k' / 'u.data'
            df = pd.read_csv(
                ratings_file, 
                sep='\t', 
                names=['user_id', 'item_id', 'rating', 'timestamp']
            )
        elif self.dataset == 'ml-1m':
            ratings_file = raw_dir / 'ml-1m' / 'ratings.dat'
            df = pd.read_csv(
                ratings_file,
                sep='::',
                names=['user_id', 'item_id', 'rating', 'timestamp'],
                engine='python'
            )
        
        # Filter implicit feedback (rating >= 4)
        df = df[df['rating'] >= 4]
        
        # Create train/val/test splits
        self._create_splits(df, processed_dir)
    
    def _create_splits(self, df: pd.DataFrame, processed_dir: Path):
        """Create temporal train/validation/test splits"""
        
        # Sort by timestamp
        df = df.sort_values('timestamp')
        
        # Split ratios
        train_ratio, val_ratio = 0.8, 0.1
        n_interactions = len(df)
        
        train_end = int(n_interactions * train_ratio)
        val_end = int(n_interactions * (train_ratio + val_ratio))
        
        # Create splits
        train_df = df.iloc[:train_end]
        val_df = df.iloc[train_end:val_end]
        test_df = df.iloc[val_end:]
        
        # Save splits
        train_df.to_csv(processed_dir / 'train.csv', index=False)
        val_df.to_csv(processed_dir / 'val.csv', index=False)
        test_df.to_csv(processed_dir / 'test.csv', index=False)
        
        print(f"Created splits - Train: {len(train_df)}, Val: {len(val_df)}, Test: {len(test_df)}")
```

### Data Preprocessing Utilities
```python
class DataPreprocessor:
    """Preprocessing utilities for recommendation data"""
    
    @staticmethod
    def filter_by_frequency(
        df: pd.DataFrame,
        min_user_interactions: int = 5,
        min_item_interactions: int = 5
    ) -> pd.DataFrame:
        """Filter users and items by minimum interaction frequency"""
        
        # Filter users
        user_counts = df['user_id'].value_counts()
        valid_users = user_counts[user_counts >= min_user_interactions].index
        df = df[df['user_id'].isin(valid_users)]
        
        # Filter items
        item_counts = df['item_id'].value_counts()
        valid_items = item_counts[item_counts >= min_item_interactions].index
        df = df[df['item_id'].isin(valid_items)]
        
        return df
    
    @staticmethod
    def remap_ids(df: pd.DataFrame) -> Tuple[pd.DataFrame, Dict[str, Dict]]:
        """Remap user and item IDs to consecutive integers"""
        
        # Create mappings
        unique_users = sorted(df['user_id'].unique())
        unique_items = sorted(df['item_id'].unique())
        
        user_mapping = {old_id: new_id for new_id, old_id in enumerate(unique_users)}
        item_mapping = {old_id: new_id for new_id, old_id in enumerate(unique_items)}
        
        # Apply mappings
        df_mapped = df.copy()
        df_mapped['user_id'] = df_mapped['user_id'].map(user_mapping)
        df_mapped['item_id'] = df_mapped['item_id'].map(item_mapping)
        
        mappings = {
            'user_mapping': user_mapping,
            'item_mapping': item_mapping,
            'reverse_user_mapping': {v: k for k, v in user_mapping.items()},
            'reverse_item_mapping': {v: k for k, v in item_mapping.items()}
        }
        
        return df_mapped, mappings
```

### Functional Requirements
- [ ] SequentialDataset handles variable-length sequences correctly
- [ ] Efficient padding and batching for sequences up to 1000 items
- [ ] Negative sampling implementations for implicit feedback
- [ ] MovieLens dataset integration with automatic download
- [ ] Train/validation/test splitting with proper temporal ordering

### Performance Requirements
- [ ] DataLoader achieves >80% GPU utilization during training
- [ ] Memory usage remains reasonable for large datasets (>1M interactions)
- [ ] Loading time <30 seconds for MovieLens-1M dataset
- [ ] Batch processing handles sequences efficiently

### Code Quality
- [ ] Full type hints and comprehensive docstrings
- [ ] Robust error handling for malformed data
- [ ] Configuration validation and sensible defaults
- [ ] Modular design for easy extension

### Unit Tests
- Test sequence padding and truncation
- Validate negative sampling distributions  
- Check vocabulary building and encoding
- Test data loading and preprocessing utilities

### Integration Tests
- End-to-end dataset loading and batch creation
- MovieLens download and preprocessing
- DataLoader integration with PyTorch training loop
- Memory usage profiling with large sequences

### Performance Tests
- Benchmark data loading speed
- Memory usage analysis
- GPU utilization during training

### Memory Optimization
- Use efficient data structures for large vocabularies
- Implement lazy loading for very large datasets
- Consider memory mapping for datasets that don't fit in RAM

### Extensibility
- Abstract base classes for custom datasets
- Plugin system for preprocessing transforms
- Configuration-driven negative sampling strategies

## Definition of Done
- All acceptance criteria met
- Unit tests passing with >90% coverage
- Integration with base model interface verified
- Performance benchmarks meet requirements
- Documentation complete with usage examples
- Ready for model training implementation

This data infrastructure provides the foundation for efficient training of sequential recommendation models while supporting the framework's scalability and performance requirements.
